src/vs/workbench/services/remote/electron-sandbox/tunnelServiceImpl.ts

```diff
@@ -0,0 +1,97 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ILogService } from 'vs/platform/log/common/log';
+import { IWorkbenchEnvironmentService } from 'vs/workbench/services/environment/common/environmentService';
+import { URI } from 'vs/base/common/uri';
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { ITunnelService, AbstractTunnelService, RemoteTunnel, TunnelPrivacyId } from 'vs/platform/remote/common/tunnel';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { IAddressProvider } from 'vs/platform/remote/common/remoteAgentConnection';
+import { ISharedProcessTunnelService } from 'vs/platform/remote/common/sharedProcessTunnelService';
+import { ILifecycleService } from 'vs/workbench/services/lifecycle/common/lifecycle';
+
+class SharedProcessTunnel extends Disposable implements RemoteTunnel {
+
+	public readonly privacy = TunnelPrivacyId.Private;
+	public readonly protocol: string | undefined = undefined;
+
+	constructor(
+		private readonly _sharedProcessTunnelService: ISharedProcessTunnelService,
+		private readonly _id: string,
+		public readonly tunnelRemoteHost: string,
+		public readonly tunnelRemotePort: number,
+		public readonly tunnelLocalPort: number | undefined,
+		public readonly localAddress: string,
+		private readonly _onBeforeDispose: () => void
+	) {
+		super();
+	}
+
+	public override async dispose(): Promise<void> {
+		this._onBeforeDispose();
+		super.dispose();
+		await this._sharedProcessTunnelService.destroyTunnel(this._id);
+	}
+}
+
+export class TunnelService extends AbstractTunnelService {
+
+	private readonly _activeSharedProcessTunnels = new Set<string>();
+
+	public constructor(
+		@ILogService logService: ILogService,
+		@IWorkbenchEnvironmentService private readonly _environmentService: IWorkbenchEnvironmentService,
+		@ISharedProcessTunnelService private readonly _sharedProcessTunnelService: ISharedProcessTunnelService,
+		@ILifecycleService lifecycleService: ILifecycleService,
+	) {
+		super(logService);
+
+		// Destroy any shared process tunnels that might still be active
+		lifecycleService.onDidShutdown(() => {
```

Comment:
Does this mean that if you have more than one VS Code window open that all your tunnels will be destroyed if you close one window?

Reply:
No, it calls `destroyTunnel` only with the `ids` that the current window has as active tunnels.